### PR TITLE
re: Replace `Arc<Store>` with `Store`

### LINKED ISSUE #4837: refactor: replace `Arc<Store>` with `Store` throughout the codebase?
We use `Arc<Store>` a bunch. But the Store is defined as 

https://github.com/near/nearcore/blob/3080cf7d305c8e1d25d09d378c7f0ab1f841ec27/core/store/src/lib.rs#L48-L52

Ie, it is already an `Arc`. I don't think there can be a conceivable reason for a double indirection.

*However*, I see that `#[derive(Clone)]` was a late addition: 

https://github.com/near/nearcore/pull/3310/files#diff-bc42e726bca3d291bcb277ae2857c8dd36ae5831d1190902b12082d74bc6e302R47

So maybe it's that derive that is wrong? Looking at the code, it seems that Clone is fine:

* there are no store methods which require exclusive access 
* `create_store` already creates an `Arc`, so we never even *have* exclusive access to the store. 

So I suggest replacing all `Arc<Store>` with just `Store`. While we are at it, we have a bunch of `get_store(&self) -> Arc<Store>` methods. I think a more idiomatic signature for those would be `store(&self) -> &Store`.

### PR DESCRIPTION
**Pull request by pmnoxx**
I was always curious, why we need `Pin` inside `Store`. It looks like we don't, and we can change
```
#[derive(Clone)]
pub struct Store {
    storage: Pin<Arc<dyn Database>>,
}
```
to
```
#[derive(Clone)]
pub struct Store {
    storage: Arc<dyn Database>,
}
```

Closes #4837

### CODE REVIEW

File: core/store/src/lib.rs [hunk 1]
```diff
@@ -44,11 +43,11 @@ mod trie;
 
 #[derive(Clone)]
 pub struct Store {
-    storage: Pin<Arc<dyn Database>>,
```
Comment (matklad) [inline core/store/src/lib.rs:47]: If we remove `Pin` here, we should remove `PhantomPinned` from `RocksDB` struct as well. 

I *think* these are fine to remove, but what gives me a pause is that I don't understand the reasonong behind their original addition: https://github.com/near/nearcore/pull/3025/files#r772310668.

Comment (pmnoxx) [reply core/store/src/lib.rs:47]: @bowenwang1996 I see, you reviewed [#3025](https://github.com/near/nearcore/pull/3025), could you tel us more about the reason why  `PhantomPinned` was added?

File: chain/chain/src/store.rs [hunk 2]
```diff
@@ -390,7 +389,7 @@ impl ChainStore {
         }
     }
 
-    pub fn owned_store(&self) -> Arc<Store> {
+    pub fn owned_store(&self) -> Store {
```
Comment (matklad) [inline chain/chain/src/store.rs:392]: This feels redundant now, as one can `.store().clone()`

File: core/store/src/lib.rs [hunk 3]
```diff
@@ -44,11 +44,11 @@ mod trie;
 
 #[derive(Clone)]
```
Comment (matklad) [inline core/store/src/lib.rs:45]: @mm-near this feels like an important type to document. I am especially wondering about our threading strategy here -- this type is `Clone`, so many threads may access the database at the same time, and they can also modify it concurrently via `Store::store_update`. And then there are things like iterators which, at least judging from the types, can happen concurrently with modification.

**Review by matklad — APPROVED** [review]

**Review by nikurt — APPROVED** [review]

File: chain/chain/src/store.rs (hunk 2 above)
Comment (pmnoxx) [reply chain/chain/src/store.rs]: Fixed.

**Comment by pmnoxx** [comment]
```
@mm-near: /chain/network/ /chain/ /nearcore/
@longarithm: /core/store/src/trie
```
@mm-near @longarithm Can you take a look?

**Comment by pmnoxx** [comment]
```
@bowenwang1996 @mm-near: /chain/ /chain/network/ /nearcore/
@longarithm @mzhangmzz: /core/store/src/trie
```
@bowenwang1996 @longarithm @mm-near @mzhangmzz Can you take a look?

**Review by Longarithm — APPROVED** [review]

**Comment by pmnoxx** [comment]
```
@longarithm @mzhangmzz: core/store/src/trie
```
@longarithm @mzhangmzz Can you take a look?

File: core/store/src/lib.rs (hunk 3 above)
Comment (mzhangmzz) [reply core/store/src/lib.rs:45]: I can comment on how store is used in the chain and runtime level and @pmnoxx can comment on how it is accessed in `PeerManagerActor` and other network related threads. My guess is that the network threads access completely different database columns than chain and runtime would access, so they don't conflict with each other.

The general story is that although `Store` can be modified concurrently, in practice we don't do that. In chain and runtime, `Store` is used by three structs, `Chain`, `NightshadeRuntime` and `EpochManager`. `Chain` is accessed by one thread (`ClientActor`) while `NightshadeRuntime` and `EpochManager` are accessed by multiple threads (`ViewClientActor`, `ClientActor` and temporary threads when we apply blocks). The trick is, only `Chain` actually modifies store through `store_update`, `EpochManager` and `NightshadeRuntime` only use `Store` to read information, they never write (except for maybe during initialization).

**Review by mzhangmzz — APPROVED** [review]

**Pull request merged**